mailet/api/src/main/java/org/apache/mailet/LoopPrevention.java

```diff
@@ -0,0 +1,70 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.mailet;
+
+import java.util.Collection;
+import java.util.Set;
+
+import org.apache.james.core.MailAddress;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+public class LoopPrevention {
+    public static final AttributeName RECORDED_RECIPIENTS_ATTRIBUTE_NAME = AttributeName.of("loop.prevention.recorded.recipients");
+
+    public static Set<MailAddress> nonRecordedRecipients(Set<MailAddress> recipients, Set<MailAddress> recordedRecipients) {
+        return Sets.difference(recipients, recordedRecipients);
+    }
+
+    public static void recordRecipients(Mail mail,
+                                        Set<MailAddress> recordedRecipients,
+                                        Set<MailAddress> newRecipients,
+                                        MailAddress originalRecipient) {
+        mail.setAttribute(createAttribute(ImmutableSet.<MailAddress>builder()
+            .addAll(recordedRecipients)
+            .addAll(newRecipients)
+            .add(originalRecipient)
+            .build()));
+    }
+
+    public static Set<MailAddress> retrieveRecordedRecipients(Mail mail) {
```

Comment:
Could be better to return a real object instead of a set (which will be a wrapper around the current `set`)
```java
public static RecordedRecipients retrieveRecordedRecipients(Mail mail) {
...
```

then you can remove previous method:
```
public static Set<MailAddress> nonRecordedRecipients(Set<MailAddress> recipients, RecordedRecipients recordedRecipients)
```

and add it in the newly created class.

That way it will prevent you from confusion on calling `nonRecordedRecipients` between the two parameters,
plus it will give us the state of the current processing rather than calling static methods in mailet processing.

Reply:
Thank for your suggestion. However, I do not really see that applying your idea would make my code better. To distinguish between two parameters, we can check parameter names (recipients, recordedRecipients).

Reply:
Well if I understand what @aduprat  suggests it is just applying some strong typing here, and moving some of the computations into this new type.

Reply:
I did give a shot at it.

Reply:
I got my idea @chibenwa .
Thank you for your modifications. 👏

Reply:
Looks better indeed ;-)